File: android_app/src/main/java/me/timeto/app/ui/home/HomeScreen.kt
package me.timeto.app.ui.home

import androidx.activity.compose.LocalActivity
import androidx.compose.foundation.background
import androidx.compose.foundation.clickable
import androidx.compose.foundation.layout.PaddingValues
import androidx.compose.foundation.layout.fillMaxHeight
import androidx.compose.foundation.layout.fillMaxSize
import androidx.compose.foundation.layout.fillMaxWidth
import androidx.compose.foundation.layout.height
import androidx.compose.foundation.layout.offset
import androidx.compose.foundation.layout.padding
import androidx.compose.foundation.lazy.rememberLazyListState
import androidx.compose.material.Text
import androidx.compose.runtime.Composable
import androidx.compose.ui.Alignment
import androidx.compose.ui.Modifier
import androidx.compose.ui.draw.clip
import androidx.compose.ui.layout.onGloballyPositioned
import androidx.compose.ui.text.font.FontWeight
import androidx.compose.ui.unit.dp
import androidx.compose.ui.unit.sp
import androidx.compose.ui.zIndex
import me.timeto.app.ui.HStack
import me.timeto.app.ui.H_PADDING
import me.timeto.app.MainActivity
import me.timeto.app.ui.VStack
import me.timeto.app.ui.ZStack
import me.timeto.app.ui.c
import me.timeto.app.ui.onePx
import me.timeto.app.ui.pxToDp
import me.timeto.app.ui.rememberVm
import me.timeto.app.ui.roundedShape
import me.timeto.app.misc.extensions.toColor
import me.timeto.app.ui.checklists.ChecklistView
import me.timeto.app.ui.Padding
import me.timeto.app.ui.SpacerW1
import me.timeto.app.ui.navigation.LocalNavigationFs
import me.timeto.app.ui.privacy.PrivacyFs
import me.timeto.app.ui.readme.ReadmeFs
import me.timeto.app.ui.whats_new.WhatsNewFs
import me.timeto.shared.ui.home.HomeVm

val HomeScreen__primaryFontSize = 16.sp

val HomeScreen__itemHeight = 36.dp
val HomeScreen__itemCircleHPadding = 6.dp
val HomeScreen__itemCircleHeight = 22.dp
val HomeScreen__itemCircleFontSize = 13.sp
val HomeScreen__itemCircleFontWeight = FontWeight.SemiBold

@Composable
fun HomeScreen() {

    val navigationFs = LocalNavigationFs.current
    val mainActivity = LocalActivity.current as MainActivity

    val (vm, state) = rememberVm {
        HomeVm()
    }

    val checklistDb = state.checklistDb
    VStack(
        modifier = Modifier
            .fillMaxSize()
            .background(c.black)
            .padding(top = mainActivity.statusBarHeightDp),
        horizontalAlignment = Alignment.CenterHorizontally,
    ) {

        HomeTimerView(
            vm = vm,
            state = state,
        )

        HomeExtraTriggersView(
            extraTriggers = state.extraTriggers,
            modifier = Modifier.padding(top = 10.dp),
            contentPadding = PaddingValues(horizontal = 50.dp),
        )

        val readmeMessage = state.readmeMessage
        if (readmeMessage != null) {
            MessageButton(
                title = readmeMessage,
                onClick = {
                    vm.onReadmeOpen()
                    navigationFs.push {
                        ReadmeFs()
                    }
                },
            )
        }

        val fdroidMessage = state.fdroidMessage
        if (fdroidMessage != null) {
            MessageButton(
                title = fdroidMessage,
                onClick = {
                    navigationFs.push {
                        PrivacyFs(isFdroid = true)
                    }
                },
            )
        }

        val whatsNewMessage = state.whatsNewMessage
        if (whatsNewMessage != null) {
            MessageButton(
                title = whatsNewMessage,
                onClick = {
                    navigationFs.push {
                        WhatsNewFs()
                    }
                },
            )
        }

        VStack(
            modifier = Modifier
                .zIndex(1f),
            horizontalAlignment = Alignment.CenterHorizontally,
        ) {

            //
            // Checklist + Main Tasks

            VStack(
                modifier = Modifier
                    .weight(1f)
                    .onGloballyPositioned { coords ->
                        val totalHeight = coords.size.height
                        vm.upListsContainerSize(
                            totalHeight = pxToDp(totalHeight),
                            itemHeight = HomeScreen__itemHeight.value,
                        )
                    },
            ) {

                val checklistScrollState = rememberLazyListState()

                val isMainTasksExists = state.mainTasks.isNotEmpty()
                val listSizes = state.listsSizes

                if (checklistDb != null) {
                    ChecklistView(
                        checklistDb = checklistDb,
                        modifier = Modifier
                            .height(listSizes.checklist.dp),
                        scrollState = checklistScrollState,
                        maxLines = 1,
                        withAddButton = false,
                        topPadding = 0.dp,
                        bottomPadding = 0.dp,
                        withNavigationPadding = false,
                    )
                }

                if (isMainTasksExists) {
                    HomeTasksView(
                        tasks = state.mainTasks,
                        modifier = Modifier
                            .height(listSizes.mainTasks.dp),
                    )
                }

                if (!isMainTasksExists && checklistDb == null)
                    SpacerW1()
            }

            state.goalBarsUi.forEach { goalBarUi ->

                HStack(
                    modifier = Modifier
                        .offset(y = 1.dp)
                        .height(HomeScreen__itemHeight),
                    verticalAlignment = Alignment.CenterVertically,
                ) {

                    ZStack(
                        modifier = Modifier
                            .padding(horizontal = H_PADDING)
                            .height(HomeScreen__itemCircleHeight)
                            .fillMaxWidth()
                            .clip(roundedShape)
                            .background(c.homeFg)
                            .clickable {
                                goalBarUi.startInterval()
                            },
                    ) {

                        ZStack(
                            modifier = Modifier
                                .fillMaxHeight()
                                .fillMaxWidth(goalBarUi.ratio)
                                .background(goalBarUi.bgColor.toColor())
                                .clip(roundedShape)
                                .align(Alignment.CenterStart),
                        )

                        Text(
                            text = goalBarUi.textLeft,
                            modifier = Modifier
                                .padding(start = HomeScreen__itemCircleHPadding, top = onePx)
                                .align(Alignment.CenterStart),
                            color = c.white,
                            fontSize = HomeScreen__itemCircleFontSize,
                            fontWeight = HomeScreen__itemCircleFontWeight,
                            lineHeight = 18.sp,
                        )

                        Text(
                            text = goalBarUi.textRight,
                            modifier = Modifier
                                .padding(end = HomeScreen__itemCircleHPadding, top = onePx)
                                .align(Alignment.CenterEnd),
                            color = c.white,
                            fontSize = HomeScreen__itemCircleFontSize,
                            fontWeight = HomeScreen__itemCircleFontWeight,
                            lineHeight = 18.sp,
                        )
                    }
                }
            }

            Padding(vertical = 8.dp)
        }
    }
}

///

@Composable
private fun MessageButton(
    title: String,
    onClick: () -> Unit,
) {
    Text(
        text = title,
        modifier = Modifier
            .padding(top = 12.dp)
            .clip(roundedShape)
            .clickable {
                onClick()
            }
            .background(c.red)
            .padding(horizontal = 10.dp)
            .padding(vertical = 4.dp),
        color = c.white,
        fontSize = 14.sp,
    )
}

File: ios_app/iphone/UI/Home/HomeScreen.swift
import SwiftUI
import shared

let HomeScreen__primaryFontSize: CGFloat = 18

let HomeScreen__hPadding: CGFloat = 8
let HomeScreen__itemHeight: CGFloat = 38
let HomeScreen__itemCircleHPadding: CGFloat = 7
let HomeScreen__itemCircleHeight: CGFloat = 24
let HomeScreen__itemCircleFontSize: CGFloat = 15
let HomeScreen__itemCircleFontWeight: Font.Weight = .semibold
let HomeScreen__itemCircleMarginTrailing: CGFloat = 8
let HomeScreen__secondaryColor: Color = MainTabsVm.companion.menuSecondaryColorDark.toColor()

struct HomeScreen: View {
    
    var body: some View {
        VmView({
            HomeVm()
        }) { vm, state in
            HomeScreenInner(
                vm: vm,
                state: state
            )
        }
    }
}

///

private struct HomeScreenInner: View {
    
    let vm: HomeVm
    let state: HomeVm.State
    
    ///
    
    @Environment(Navigation.self) private var navigation
    
    var body: some View {
        
        VStack {
            
            let checklistDb: ChecklistDb? = state.checklistDb
            
            HomeTimerView(vm: vm, state: state)
            
            if let whatsNewMessage = state.whatsNewMessage {
                Button(
                    action: {
                        navigation.push(.whatsNew)
                    },
                    label: {
                        Text(whatsNewMessage)
                            .foregroundColor(.white)
                            .padding(.horizontal, 12)
                            .padding(.vertical, 8)
                            .font(.system(size: 17, weight: .medium))
                            .background(roundedShape.fill(.red))
                            .padding(.top, 8)
                    }
                )
            }
            
            let isMainTasksExists = !state.mainTasks.isEmpty
            
            GeometryReader { geometry in
                
                let _ = vm.upListsContainerSize(
                    totalHeight: Float(geometry.size.height),
                    itemHeight: Float(HomeScreen__itemHeight)
                )
                
                VStack {
                    
                    if let checklistDb = checklistDb {
                        VStack {
                            ChecklistView(
                                checklistDb: checklistDb,
                                maxLines: 1,
                                withAddButton: false,
                                onDelete: {}
                            )
                        }
                        .frame(height: CGFloat(state.listsSizes.checklist))
                    }
                    
                    if isMainTasksExists {
                        HomeTasksView(
                            tasks: state.mainTasks
                        )
                        .frame(height: CGFloat(state.listsSizes.mainTasks))
                    }
                    
                    Spacer()
                }
            }
            
            if state.showReadme {
                HomeReadmeView(
                    title: state.readmeTitle,
                    buttonText: state.readmeButtonText,
                    onButtonClick: {
                        vm.onReadmeOpen()
                    }
                )
            }
            
            HomeButtonsView()
            
            Padding(vertical: 10.0)
        }
        .padding(.bottom, MainTabsView__HEIGHT)
    }
}
